### PR TITLE
Fix skipping last segment logic in ROS.TryGet

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/dotnet/corefx/issues/28251

### CODE REVIEW

**Comment by pakrym** [comment]
I'm not sure that we want all the skipping logic altogether.

**Comment by pakrym** [comment]
Removed skipping logic from multi-segment code path. Inlined method.

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 1]
```diff
@@ -163,122 +163,165 @@ public void SeekSkipsEmptySegments()
         }
 
         [Fact]
-        public void SeekEmptySkipDoesNotCrossPastEnd()
+        public void TryGetReturnsEmptySegments()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: It's difficult to review (in the diff) what tests are changing here. Are we modifying any existing tests here or just adding some (and git diff is getting confused)?

Comment (davidfowl) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: We changed the behavior so old tests are invalid and replaced with new ones that show the new behavior.

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 2]
```diff
@@ -163,122 +163,165 @@ public void SeekSkipsEmptySegments()
...
-            Assert.Equal(bufferSegment2, c1.GetObject());
+            var start = buffer.Start;
+            Assert.True(buffer.TryGet(ref start, out var memory));
+            Assert.Equal(100, memory.Length);
+            Assert.False(buffer.TryGet(ref start, out memory));
+        }
 
-            c1 = buffer.GetPosition(100, buffer.Start);
+        [Fact]
+        public void TryGetStopsAtEndWhenEndIsFistByteOfFull()
```
Comment (davidfowl) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: nit: Fist -> First

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 3]
```diff
@@ -163,122 +163,165 @@ public void SeekSkipsEmptySegments()
...
+            Assert.True(buffer.TryGet(ref start, out var memory));
+            Assert.Equal(100, memory.Length);
+            Assert.True(buffer.TryGet(ref start, out memory));
+            Assert.Equal(0, memory.Length);
+            Assert.False(buffer.TryGet(ref start, out memory));
         }
 
         [Fact]
-        public void TryGetSkipsEmptyForNextUsingGetPositionWithOffset()
+        public void TryGetStopsAtEndWhenEndIsFistByteOfEmpty()
```
Comment (davidfowl) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: -> Fist -> First

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 4]
```diff
@@ -163,122 +163,165 @@ public void SeekSkipsEmptySegments()
...
-            c1 = buffer.GetPosition(100, buffer.Start);
+        [Fact]
+        public void TryGetStopsAtEndWhenEndIsFistByteOfFull()
+        {
+            var bufferSegment1 = new BufferSegment<byte>(new byte[100]);
+            BufferSegment<byte> bufferSegment2 = bufferSegment1.Append(new byte[100]);
 
-            Assert.Equal(0, c1.GetInteger());
-            Assert.Equal(bufferSegment2, c1.GetObject());
+            var buffer = new ReadOnlySequence<byte>(bufferSegment1, 0, bufferSegment2, 0);
```
Comment (davidfowl) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: This shouldn't affect the buffer reader right? We cover all of these cases in our tests yeh?

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: I see. We are only modifying the tests involving TryGet here, correct?

Are test cases like these no longer valid?
```C#
        [Fact]
        public void TryGetSkipDoesNotCrossPastEndUsingGetPositionWithOffsetAndOrigin()
        {
            var bufferSegment1 = new BufferSegment<byte>(new byte[100]);
            BufferSegment<byte> bufferSegment2 = bufferSegment1.Append(new byte[0]);
            BufferSegment<byte> bufferSegment3 = bufferSegment2.Append(new byte[0]);
            BufferSegment<byte> bufferSegment4 = bufferSegment3.Append(new byte[100]);

            var buffer = new ReadOnlySequence<byte>(bufferSegment1, 0, bufferSegment2, 0);
            SequencePosition c1 = buffer.GetPosition(0, buffer.Start);
            Assert.Equal(0, c1.GetInteger());
            Assert.Equal(bufferSegment1, c1.GetObject());

            ReadOnlyMemory<byte> data;
            Assert.True(buffer.TryGet(ref c1, out data, advance: true));

            Assert.Equal(0, c1.GetInteger());
            Assert.Equal(null, c1.GetObject());

            Assert.False(buffer.TryGet(ref c1, out data, advance: true));
        }
```

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 5]
```diff
@@ -163,122 +163,165 @@ public void SeekSkipsEmptySegments()
...
+            }
 
-            Assert.False(buffer.TryGet(ref c1, out data, advance: true));
+            Assert.Equal(1, sizes.Count);
+            Assert.Equal(new [] { 100 }, sizes);
         }
 
         [Fact]
-        public void TryGetSkipsEmptyForNextUsingGetPositionWithOffsetAndOrigin()
+        public void EnumerableStopsAtEndWhenEndIsFistByteOfFull()
```
Comment (davidfowl) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Fist -> First, did you mean first segment?

Comment (pakrym) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: First Byte of Empty [Segment]

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs (hunk 1 above)
Comment (pakrym) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: No, they were checking that TryGet skips empty segments which it doesn't do anymore

Comment (benaadams) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Except the tests were testing `GetPosition` which uses `Seek` not `TryGet` and you've not changed `Seek`, so  `Seek`/`GetPosition` now has less tests

Comment (pakrym) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: K, I'll bring them back

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 6]
```diff
@@ -163,122 +163,165 @@ public void SeekSkipsEmptySegments()
...
+            }
 
-            Assert.False(buffer.TryGet(ref c1, out data, advance: true));
+            Assert.Equal(2, sizes.Count);
+            Assert.Equal(new [] { 100, 0 }, sizes);
         }
 
         [Fact]
-        public void TryGetSkipDoesNotCrossPastEndUsingGetPositionWithOffsetAndOrigin()
+        public void SeekEmptySkipDoesNotCrossPastEnd()
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Ah has moved here?

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs (hunk 1 above)
Comment (benaadams) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: You just swapped where the test was :)

Comment (benaadams) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Is ok, you just swapped where the test was :) 

(According to GitHub's bad diffing)

Comment (pakrym) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Removed tests were only doing `GetPosition(0)`

Comment (benaadams) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Was specifically concerned about `SeekEmptySkipDoesNotCrossPastEnd`; but its still there, just "moved"

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs (hunk 4 above)
Comment (davidfowl) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: I see

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Pavel, how has code coverage been affected by these test changes?

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 7]
```diff
@@ -163,122 +163,165 @@ public void SeekSkipsEmptySegments()
...
+            Assert.True(buffer.TryGet(ref start, out memory));
+            Assert.Equal(100, memory.Length);
+            Assert.True(buffer.TryGet(ref start, out memory));
+            Assert.Equal(100, memory.Length);
+            Assert.False(buffer.TryGet(ref start, out memory));
+        }
 
-            var buffer = new ReadOnlySequence<byte>(bufferSegment1, 0, bufferSegment2, 0);
+        [Fact]
+        public void TryGetStopsAtEndWhenEndIsLastByteOfFull()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: `TryGetStopsAtEndWhenEndIsLastByteOfFull` - I don't understand what it means by LastByteOfFull. Can you please elaborate?

Comment (pakrym) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: LastByteOfFull[Segment], so end pointer is last byte of segment with data

**Comment by pakrym** [comment]
@dotnet-bot test OSX x64 Debug Build  please

**Comment by ahsonkhan** [comment]
@dotnet-bot test Linux x64 Release Build

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs (hunk 1 above)
Comment (pakrym) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Didn't change at all, these are commonly used code paths.

**Comment by benaadams** [comment]
Rebase?

**Comment by pakrym** [comment]
@dotnet-bot test Windows x86 Release Build

**Pull request merged**